### ISSUE TITLE
Make Datagrams return BytesMut, not Bytes

### ISSUE THREAD

**Issue by tbraun96** [quinn-rs/quinn#1173]
When zero-copy is the goal AND packet payloads need to be decrypted through AEAD, having a packet that is mutable is preferred. When Datagrams returns ``Bytes``, we must manually copy all the bytes into a separate ``BytesMut`` in memory; this is not good for throughput.

**Comment by Ralith** [comment]
> packet payloads need to be decrypted through in-place AEAD

QUIC already handles encryption for you. What are you trying to do?

**Comment by tbraun96** [comment]
I know. There is a custom protocol above QUIC for special clients that is beyond public-domain cryptography.

Even if this wasn't for cryptographic purposes, having a mutable buffer is useful for mutating data.

**Comment by Ralith** [comment]
Yeah, I don't see a drawback.

**Comment by tbraun96** [comment]
Okay, added a PR #1182

**Comment by TheButlah** [comment]
This would also be useful for me. `tokio_serde::Framed` expects a `BytesMut` in order to function, which means I need to unecessarily clone the data returned by quinn. And later, I expect to want to do in-place mutation on data, which would also necessitate a BytesMut.

Is there any fundamental reason why quinn couldn't return a BytesMut to the user for datagrams?

**Comment by djc** [comment]
@TheButlah have you already looked over the discussion in #1182?

**Comment by TheButlah** [comment]
Yes, but I am commenting from the perspective of a user, rather than someone with knowledge of the internals of quinn

**Comment by djc** [comment]
I think solving this issue will likely depend on a user pitching in as a contributor. Also we just did a semver-breaking release so if you want this sooner we might need to add an alternate method for this.
